src/yggdrasil/packetconn.go

```diff
@@ -0,0 +1,277 @@
+		sessionMTU := session._getMTU()
+		if types.MTU(len(b)) > sessionMTU {
+			sendErr <- PacketConnError{maxsize: int(sessionMTU)}
+			return
+		}
+
+		// If we got to this point then our initial checks passed - there
+		// isn't much point blocking the sender any further so release it.
+		sendErr <- nil
+
+		// Send the packet.
+		session._send(msg)
+
+		// Session keep-alive, while we wait for the crypto workers from send
+		switch {
+		case time.Since(session.time) > 6*time.Second:
+			if session.time.Before(session.pingTime) && time.Since(session.pingTime) > 6*time.Second {
+				// TODO double check that the above condition is correct
+				c.sessions.router.Act(session, func() {
+					// Check to see if there is a search already matching the
+					// supplied destination. If there is then don't start another
+					// one.
+					sinfo, isIn := c.sessions.router.searches.searches[*nodeID]
+					if !isIn {
+						// Nothing was found, so create a new search.
+						searchCompleted := func(sinfo *sessionInfo, e error) {}
+						sinfo = c.sessions.router.searches.newIterSearch(nodeID, nodeMask, searchCompleted)
+
+						// Start the search.
+						sinfo.startSearch()
+						c.sessions.router.core.log.Debugf("DHT search started: %p", sinfo)
+					}
+				})
+			} else {
+				session._sendPingPong(false)
+			}
+
+		case session.reset && session.pingTime.Before(session.time):
+			session._sendPingPong(false)
+
+		default:
+		}
+	})
+
+	// Wait for the checks to pass. Then return the success
+	// values to the caller.
+	err = <-sendErr
+	return len(b), err
+}
+
+// implements net.PacketConn
+func (c *PacketConn) Close() error {
+	// TODO: implement this. don't know what makes sense for net.PacketConn?
+	return nil
+}
+
+// implements net.PacketConn
+func (c *PacketConn) LocalAddr() net.Addr {
+	return &c.sessions.router.core.boxPub
+}
+
+// SetReadCallback allows you to specify a function that will be called whenever
+// a packet is received. This should be used if you wish to implement
+// asynchronous patterns for receiving data from the remote node.
+//
+// Note that if a read callback has been supplied, you should no longer attempt
+// to use the synchronous Read function.
+func (c *PacketConn) SetReadCallback(callback func(net.Addr, []byte)) {
+	c.Act(nil, func() {
+		c.readCallback = callback
+		c._drainReadBuffer()
+	})
+}
+
+func (c *PacketConn) _drainReadBuffer() {
+	if c.readCallback == nil {
+		return
+	}
+	select {
+	case bs := <-c.readBuffer:
+		c.readCallback(bs.addr, bs.payload)
+		c.Act(nil, c._drainReadBuffer) // In case there's more
+	default:
+	}
+}
+
+// SetDeadline is equivalent to calling both SetReadDeadline and
+// SetWriteDeadline with the same value, configuring the maximum amount of time
+// that synchronous Read and Write operations can block for. If no deadline is
+// configured, Read and Write operations can potentially block indefinitely.
+func (c *PacketConn) SetDeadline(t time.Time) error {
```

Comment:
As far as I can see, it looks like the `Deadline`s aren't currently implemented. We should probably document that for now. We probably need to think about how we can implement them and get the correct behavior. I can't think of a good way to do it (i.e. return in a bounded amount of time after the deadline without leaking any goroutines or something) without resorting to timers and `select`ing on channels. Which is OK for the blocking `PacketConn` interface I guess, and maybe an actor-based interface can just bypass it entirely with async message sends, but that seems a little ugly.